### ISSUE TITLE
Produce a .dot file from recursive dependencies?

### ISSUE THREAD

**Issue by angerman** [ndmitchell/shake#552]
From time to time I run into the situation where I get
```
Build system error - indirect recursion detected:
  Key value 1:    start
  ...
  Key value 111: ...
Rules may not be recursive
```

It would be really helpful if shake could produce a graph. Say key value numbers as labels?
That would make understanding the issue much easier I believe. The actual dependencies
are likely too large to put into labels. As such the keys would have to be enough and one
would have to refer to the keys.

**Comment by ndmitchell** [comment]
Is this a bug when you are developing and have introduced a problem? If you run with -j1 do you get a better error with the exact smaller cycle?

The problem is you’re hitting a corner case in the recursion detection algorithm and Shake has no idea what the graph is - it’s encoded in opaque IO values it can’t peer into.

**Comment by angerman** [comment]
Yes, this is a bug, I introduced in my rules. I *can* read the output, and work through it by hand. But if shake would just provide a dot file it would be faster.

And the information is there (has been so far), on the right hand side of the key.
So what I would envision is that shake just remembers the (key, value) pairs, and builds a graph from them, based on the equality of the rhs value.

Maybe even simpler if it knew the the (key, value) list before printing the system error message,
add
```
.-  Key value 1:  XXX
|   Key value 2: ...
|   ...
|   Key value N: ...
'-  Key value (N+1): XXX
```

in front of the keys, where the values match? But then again maybe this is just a symptom of the code I work with.

**Comment by ndmitchell** [comment]
Next time it occurs can you post a full trace? At this point of failure Shake doesn't have much information to hand, pretty much just what it displays on screen, but if it can help with a better rendering of that information I'll certainly give it a go.
